### PR TITLE
chore(dafny): Check for unique transformed keys on init mutation

### PR DESCRIPTION
**Pull request by rishav-karanjit**
### Description of changes:
On initialize mutation, while mutating from hv1 to hv2 fail if EC does not have unique transformed keys

### Squash/merge commit message, if applicable:

```
<type>(dafny/java/python/dotnet/go/rust): <description>
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStoreAdmin/src/InitializeMutation.dfy [hunk 1]
```diff
@@ -195,14 +196,24 @@ module {:options "/functionSyntax:4" } InternalInitializeMutation {
...
+        && input.Mutations.TerminalHierarchyVersion.Some?
+        && input.Mutations.TerminalHierarchyVersion.value.v2?
+      ) {
+      :- Need(
+        HvUtils.HasUniqueTransformedKeys?(readItems.ActiveItem.EncryptionContext),
+        Types.UnexpectedStateException(
+          message :=
+            KeyStoreErrorMessages.NOT_UNIQUE_BRANCH_KEY_CONTEXT_KEYS
+        )
+      );
```
Comment (texastony) [inline AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStoreAdmin/src/InitializeMutation.dfy:215]: This is good and is a start, 
but I believe we will need to also check
the union of the terminalEC and the inferredEC.

See lines 252 to 278;
I am still trying to imagine an edge case b/w new EC and old EC...

File: AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStoreAdmin/test/Mutations/TestHierarchyVersion.dfy [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    // Commented code that adds {"Robbie": "Is a dog."} to the dynamodb item
+    // This code will create a item that contains non unique branch key context key
+    //
+    // var ddbClient :- expect Fixtures.ProvideDDBClient();
+    // var kmsClient :- expect Fixtures.ProvideKMSClient();
+    // var _ :- expect AdminFixtures.AddAttributeWithoutLibrary(
+    //   id:="DO-NOT-EDIT-Branch-Key-For-HasUniqueTransformedKeys-Check",
+    //   keyValue:=AdminFixtures.KeyValue(key:="Robbie", value:="Is a dog."),
+    //   alsoViolateBeacon? := true, ddbClient? := Some(ddbClient),
+    //   kmsClient?:=Some(kmsClient), violateReservedAttribute:=true);
```
Comment (texastony) [inline AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStoreAdmin/test/Mutations/TestHierarchyVersion.dfy:26]: Blocking: Add the logic to create the Branch Key to the commented out block.
i.e: create the healthy BK, then violate it, all in commented out code.
There should be a helper method in Fixtures or Admin Fixtures.

File: AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStoreAdmin/src/InitializeMutation.dfy [hunk 3]
```diff
@@ -195,14 +196,24 @@ module {:options "/functionSyntax:4" } InternalInitializeMutation {
...
+    if (
+        && input.Mutations.TerminalHierarchyVersion.Some?
+        && input.Mutations.TerminalHierarchyVersion.value.v2?
+      ) {
+      :- Need(
+        HvUtils.HasUniqueTransformedKeys?(readItems.ActiveItem.EncryptionContext),
+        Types.UnexpectedStateException(
+          message :=
+            KeyStoreErrorMessages.NOT_UNIQUE_BRANCH_KEY_CONTEXT_KEYS
+        )
```
Comment (texastony) [inline AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStoreAdmin/src/InitializeMutation.dfy:214]:
```suggestion
      // TODO-HV-2-M2 : Check combination of terminalEC and inferredEC for unique EC
      :- Need(
        HvUtils.HasUniqueTransformedKeys?(readItems.ActiveItem.EncryptionContext),
        Types.UnexpectedStateException(
          message :=
            KeyStoreErrorMessages.NOT_UNIQUE_BRANCH_KEY_CONTEXT_KEYS
        )
```

Comment (texastony) [inline AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStoreAdmin/src/InitializeMutation.dfy:214]: Blocking: I think I want to call this a user error, 
as they MUST have modified the BK without using the library.
In which case, this would be KeyStoreAdminException
```suggestion
      :- Need(
        HvUtils.HasUniqueTransformedKeys?(readItems.ActiveItem.EncryptionContext),
        Types.KeyStoreAdminException(
          message :=
            KeyStoreErrorMessages.NOT_UNIQUE_BRANCH_KEY_CONTEXT_KEYS
        )
```

File: AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStoreAdmin/test/Mutations/TestHierarchyVersion.dfy [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+module {:options "/functionSyntax:4" } TestHierarchyVersion {
+  import Types = AwsCryptographyKeyStoreAdminTypes
+  import KeyStoreTypes = AwsCryptographyKeyStoreTypes
+  import KeyStoreErrorMessages
+  import AdminFixtures
+  import Fixtures
+  import opened Wrappers
+
+  method {:test} {:vcs_split_on_every_assert} TestInitializeMutationFailsWithNonUniqueBranchKeyContext() {
```
Comment (texastony) [inline AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStoreAdmin/test/Mutations/TestHierarchyVersion.dfy:16]: Blocking: Refactor test to expect `KeyStoreAdminException`.

**Review by texastony — CHANGES_REQUESTED** [review]
Good Start.
Blockers: make the error and test revisions and a TODO for a check on the combination of terminalEC and inferredEC.

We can merge with the TODO and you can pick up the TODO in the next PR.

File: AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStoreAdmin/test/Mutations/TestHierarchyVersion.dfy [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+  import Fixtures
+  import opened Wrappers
+
+  method {:test} TestInitializeMutationFailsWithNonUniqueBranchKeyContext() {
+
+    var testId := "DO-NOT-EDIT-Branch-Key-For-HasUniqueTransformedKeys-Check";
+    var ddbClient :- expect Fixtures.ProvideDDBClient();
+    var kmsClient :- expect Fixtures.ProvideKMSClient();
+    // Commented code that adds {"Robbie": "Is a dog."} to the dynamodb item "DO-NOT-EDIT-Branch-Key-For-HasUniqueTransformedKeys-Check" in table KeyStoreDdbTable
+    // This code will create a branch key and make changes so that branch key item contains non unique branch key context key
```
Comment (rishav-karanjit) [inline AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStoreAdmin/test/Mutations/TestHierarchyVersion.dfy:22]: This comment on line (21 and 22) is changed in downstream PR https://github.com/aws/aws-cryptographic-material-providers-library/pull/1411/files

**Comment by rishav-karanjit** [comment]
Closing as rebased branch is here https://github.com/aws/aws-cryptographic-material-providers-library/pull/1417
